core/src/main/java/org/elasticsearch/search/sort/SortBuilder.java

```diff
@@ -96,4 +95,122 @@ public T order(SortOrder order) {
+        if (token == XContentParser.Token.START_ARRAY) {
+            while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                if (token == XContentParser.Token.START_OBJECT) {
+                    parseCompoundSortField(parser, context, sortFields);
+                } else if (token == XContentParser.Token.VALUE_STRING) {
+                    String fieldName = parser.text();
+                    if (fieldName.equals(ScoreSortBuilder.NAME)) {
+                        sortFields.add(new ScoreSortBuilder());
+                    } else {
+                        sortFields.add(new FieldSortBuilder(fieldName));
+                    }
+                } else {
+                    throw new IllegalArgumentException("malformed sort format, "
+                            + "within the sort array, an object, or an actual string are allowed");
+                }
+            }
+        } else if (token == XContentParser.Token.VALUE_STRING) {
+            String fieldName = parser.text();
+            if (fieldName.equals(ScoreSortBuilder.NAME)) {
+                sortFields.add(new ScoreSortBuilder());
+            } else {
+                sortFields.add(new FieldSortBuilder(fieldName));
+            }
+        } else if (token == XContentParser.Token.START_OBJECT) {
+            parseCompoundSortField(parser, context, sortFields);
+        } else {
+            throw new IllegalArgumentException("malformed sort format, either start with array, object, or an actual string");
+        }
+        return sortFields;
+    }
+
+    private static void parseCompoundSortField(XContentParser parser, QueryParseContext context, List<SortBuilder<?>> sortFields)
+            throws IOException {
+        XContentParser.Token token;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                String fieldName = parser.currentName();
+                token = parser.nextToken();
+                if (token == XContentParser.Token.VALUE_STRING) {
+                    SortOrder order = SortOrder.fromString(parser.text());
+                    if (fieldName.equals(ScoreSortBuilder.NAME)) {
+                        sortFields.add(new ScoreSortBuilder().order(order));
+                    } else {
+                        sortFields.add(new FieldSortBuilder(fieldName).order(order));
+                    }
+                } else {
+                    if (PARSERS.containsKey(fieldName)) {
+                        sortFields.add(PARSERS.get(fieldName).fromXContent(context, fieldName));
+                    } else {
+                        sortFields.add(FieldSortBuilder.PROTOTYPE.fromXContent(context, fieldName));
+                    }
+                }
+            }
+        }
+    }
+
+    public static void parseSort(XContentParser parser, SearchContext context) throws IOException {
```

Comment:
Yes, this will be used to add compatibility with other users of current SortParseElement. I can take it out of this PR though.